rfcs/NNN-form-associated-mixin.md

```diff
@@ -0,0 +1,377 @@
+case.
+
+### Why decorators?
+
+A key choice in this design is to not require or add any _public_ API of the implementing element. Rather than automatically adding `.value` or `.disabled`
+fields, we leave that up to the element authors.
+
+In order for the mixin to function, it needs to perform some actions and read
+some element state when form-relate state changes. Decorators let us wrap
+arbitrary element accessors to invoke the mixin code paths to update form state,
+and let us communicate hooks to the mixin declaratively.
+
+```ts
+class MyFormElement extends FormAssociated(LitElement) {
+
+  // Setting this field calls internals.setFormValue()
+  @formValue()
+  @property({reflect: true})
+  accessor value: string = '';
+
+  // When the form resets, the value is set to this field
+  @formDefaultValue()
+  accessor defaultValue: string = '';
+
+  render() {
+    return html`<input .value=${this.value}> @input=${this.#onInput}`;
+  }
+
+  #onInput(e) {
+    this.value = e.target.value;
+  }
+}
+```
+
+Becauase decorators are used to denote the important form fields, name of the form value field does not need to be `value`. For instance, a checkbox element could
+use the field `checked`:
+
+
+```ts
+class MyCheckbox extends FormAssociated(LitElement) {
+
+  @formValue({
+    converter: {
+      toFormValue(value: boolean) {
+        return String(value);
+      },
+      fromFormValue(value: string) {
+        return Boolean(value);
+      },
+    })
+  @property({type: Boolean, reflect: true})
+  accessor checked: boolean = false;
+}
+```
+
+_Note on this example: "boolean" has to be repeated three times here. We can't
+deduplicate the TypeScript annotation and the type used by the `@property()`
+converter, but we might be able to come up with a way for `@formValue()` to
+automatically read the type from the `@property()` property options._
+
+### The FormAssociated mixin
+
+The `FormAssociated()` mixin implements most of the required logic for manging
+the form-related state on `ElementInternals`.
+
+#### formAssociated
+
+The `FormAssociated()` mixin defines `static formAssociated = true`
+
+#### ElementInternals
+
+The `FormAssociated()` mixin needs access to `ElementInternals`, so it calls `this.attachInternals()` in the constructor.
+
+Elements applying the mixin may also need access to internals, but
+`attachInternals()` is only allowed to be called once so that outside callers
+cannot access internals, so a `getInternals()` utility is provided to get them.
+
+To discourage use of `getInternals()` by outside code, it must be called with a
+reference to the class and can only be be called once per class.
+
+#### Form Value
+
+A form-associated element's form value must be passed to `internals.setValue()`
+whenever the value is changed. This is implemented by the `@formValue()`
+decorator, which wrapps an accesor's setter to call `setValue()`.
+
+A form value must be of type `string | File | FormData | null`. To facilitiate
```

Comment:
So I take that you can mimic `<select multiple>` by having a `FormData` value containing multiple values for the same key, right? Is that something to spend a few words on?

Reply:
Yes, but this sentence is merely reflecting the platform API constraints. This library won't dictate or interfere with how form data is represented, so however you would do this manually is how you would do it here.